[Flaky Test] `Shoot Maintenance controller tests`

<!-- Please only use this template for submitting reports about flaky tests or jobs (pass or fail with no underlying change in code) in Gardener CI -->

**How to categorize this issue?**
<!--
Please select area, kind, and priority for this issue. This helps the community categorizing it.
Replace below TODOs or exchange the existing identifiers with those that fit best in your opinion.
If multiple identifiers make sense you can also state the commands multiple times, e.g.
  /area control-plane
  /area auto-scaling
  ...

"/area" identifiers:     audit-logging|auto-scaling|backup|certification|control-plane-migration|control-plane|cost|delivery|dev-productivity|disaster-recovery|documentation|high-availability|logging|metering|monitoring|networking|open-source|ops-productivity|os|performance|quality|robustness|scalability|security|storage|testing|usability|user-management
"/kind" identifiers:     api-change|bug|cleanup|discussion|enhancement|epic|impediment|poc|post-mortem|question|regression|task|technical-debt|test
-->
/area testing
/kind flake

**Which test(s)/suite(s) are flaking**:
`Shoot Maintenance controller tests`

**CI link**: https://concourse.ci.gardener.cloud/teams/gardener/pipelines/gardener-master/jobs/master-head-update-job/builds/178

**Reason for failure**:

```
• [SLOW TEST:10.023 seconds]
Shoot Maintenance controller tests
/go/src/github.com/gardener/gardener/test/integration/envtest/shootmaintenance/maintenance_control_test.go:37
  Kubernetes version maintenance tests
  /go/src/github.com/gardener/gardener/test/integration/envtest/shootmaintenance/maintenance_control_test.go:202
    Kubernetes version should not be updated: auto update not enabled
    /go/src/github.com/gardener/gardener/test/integration/envtest/shootmaintenance/maintenance_control_test.go:203
------------------------------
•panic: assignment to entry in nil map

goroutine 8441 [running]:
github.com/gardener/gardener/pkg/controllerutils.setTaskAnnotations(...)
	/go/src/github.com/gardener/gardener/pkg/controllerutils/miscellaneous.go:86
github.com/gardener/gardener/pkg/controllerutils.AddTasks(0xc0006b7100, {0xc004607ba0, 0x1, 0x36})
	/go/src/github.com/gardener/gardener/pkg/controllerutils/miscellaneous.go:59 +0xc8
github.com/gardener/gardener/pkg/controllermanager/controller/shoot.(*shootMaintenanceReconciler).reconcile(0xc002d2f900, {0x2d81958, 0xc0029e62d0}, 0xc0045a3500, {0x2dcc3f0, 0xc002fcdd50})
	/go/src/github.com/gardener/gardener/pkg/controllermanager/controller/shoot/shoot_maintenance_control.go:176 +0x7a7
github.com/gardener/gardener/pkg/controllermanager/controller/shoot.(*shootMaintenanceReconciler).Reconcile(0xc002d2f900, {0x2d81958, 0xc0029e62d0}, {{{0xc002710460, 0x2728600}, {0xc002710450, 0xc007842580}}})
	/go/src/github.com/gardener/gardener/pkg/controllermanager/controller/shoot/shoot_maintenance_control.go:119 +0x128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc005e62820, {0x2d81958, 0xc0029e6240}, {{{0xc002710460, 0x2728600}, {0xc002710450, 0xc00337bf40}}})
	/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114 +0x222
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc005e62820, {0x2d818b0, 0xc00337be40}, {0x2638480, 0xc007b68820})
	/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311 +0x2f2
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc005e62820, {0x2d818b0, 0xc00337be40})
	/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:223 +0x354
FAIL	github.com/gardener/gardener/test/integration/envtest/shootmaintenance	38.940s
FAIL
make: *** [Makefile:239: test-integration] Error 1
```

It seems that there is a race on Shoot creation between the reconciliation that is enqueued by the watch for the test purposes and the reconciliation that is triggered by the annotation. If the reconciliation is triggered by the watch itself, then Shoot .metadata.annotation will be nil and the controller will panic with `assignment to entry in nil map`.
One improvement that we have to do is to adapt the Shoot maintenance controller to handle the case when Shoot .metadata.annotations=nil. We also have to remove this race in the test between the trigger from the watch and the trigger from the gardener.cloud/operation annotation.

--- a/test/integration/envtest/shootmaintenance/maintenance_control_test.go
+++ b/test/integration/envtest/shootmaintenance/maintenance_control_test.go
@@ -31,6 +31,7 @@ import (
 	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
 	"github.com/gardener/gardener/pkg/controllermanager/apis/config"
 	shootcontroller "github.com/gardener/gardener/pkg/controllermanager/controller/shoot"
+	"github.com/gardener/gardener/pkg/utils"
 	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
 )
 
@@ -93,6 +94,7 @@ var _ = Describe("Shoot Maintenance controller tests", func() {
 									Version:        testMachineImageVersion,
 									Classification: &deprecatedClassification,
 								},
+								CRI: []gardencorev1beta1.CRI{{Name: gardencorev1beta1.CRINameDocker}},
 							},
 						},
 					},
@@ -146,6 +148,10 @@ var _ = Describe("Shoot Maintenance controller tests", func() {
 					AutoUpdate: &gardencorev1beta1.MaintenanceAutoUpdate{
 						KubernetesVersion:   false,
 						MachineImageVersion: false,
+					},
+					TimeWindow: &gardencorev1beta1.MaintenanceTimeWindow{
+						Begin: utils.NewMaintenanceTime(time.Now().Add(2*time.Hour).Hour(), 0, 0).Formatted(),
+						End:   utils.NewMaintenanceTime(time.Now().Add(4*time.Hour).Hour(), 0, 0).Formatted(),
 					},
 				},
 			},

--- a/test/integration/envtest/shootmaintenance/maintenance_suite_test.go
+++ b/test/integration/envtest/shootmaintenance/maintenance_suite_test.go
@@ -37,11 +37,7 @@ import (
 	"github.com/gardener/gardener/test/framework"
 )
 
-// TODO: This test suite is disabled because of a known flake in the test (https://github.com/gardener/gardener/issues/4944).
-// Enable the test suite again after fixing the flake.
-
 func TestShootMaintenance(t *testing.T) {
-	t.Skipf("Temporarily skipping the test suite because of known flake...")
 	controllermanagerfeatures.RegisterFeatureGates()
 	RegisterFailHandler(Fail)
 	RunSpecs(t, "Shoot Maintenance Controller Integration Test Suite")

